### PR TITLE
[JENKINS-72136] Remove inline onclick handlers for generate directive button

### PR DESCRIPTION
**Pull request by NeetigyaPod**
* JENKINS issue(s):
    * https://issues.jenkins.io/browse/JENKINS-72136
* Description:
    * Inline handlers were showing in console. Used the recommendations given in the issue.
* Documentation changes:
    * There is no feature change but just a minor change for security reasons.
* Users/aliases to notify:
    * n/a
 
Testing:
Screenshots:
Before:
Screenshot from 2023-10-12 11-16-04
After:

Screenshot from 2023-10-12 11-17-57

mvn tests run and changes appear in mvn:hpi run

### CODE REVIEW

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (Kevin-CB) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:81]:
```suggestion
                    <span class="prototype-button-reference-holder" data-id="${id}" data-fullurl="${rootURL}/${it.GENERATE_URL}" />
```

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (Kevin-CB) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:82]:
```suggestion
                    <input id="prototypeButton_${id}" type="button" value="${%Generate Declarative Directive}" class="submit-button primary"/>
```

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (Kevin-CB) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:84]:
```suggestion
                    <st:adjunct includes="org.jenkinsci.plugins.pipeline.modeldefinition.generator.DirectiveGenerator.indexScript"/>
```

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js
Comment (Kevin-CB) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js:1]:
```suggestion
Behaviour.specify(".directive-generator-button-reference-holder", 'prototype', 0, function (e) {
```

**Review by Kevin-CB — COMMENTED** [review]
I tested your changes locally, with the CSP plugins, and it works fine! 
I'm not able to see the previous CSP violations!

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly]: Done

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly]: Done

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js]: Needed to change the class in jelly file as well, done

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly]: Done

**Comment by NeetigyaPod** [comment]
Hi, that's great. Sorry for the extra spaces. Will take care in future contributions.

**Comment by NeetigyaPod** [comment]
Hi @Kevin-CB  can we merge this before end of October? I would like it to be a part of my hacktoberfest contribution.

**Review by Kevin-CB — APPROVED** [review]

**Comment by Kevin-CB** [comment]
Hi @NeetigyaPod, even though I approved this PR, I can't merge it as I'm not a maintainer of this plugin.

**Comment by NeetigyaPod** [comment]
@jglick Can you please help me in this? Would appreciate a lot

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js
Comment (jglick) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js:14]: likely needs to be revisited in another PR

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (jglick) [inline pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:79]: Could probably get rid of the `generateId` trick now, by having the button look for a sibling textarea.

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/indexScript.js:14]: Can be done. Will work on it after this one

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:79]: Sibling in the sense- I can allocate a similar class name to the text area and remove all the mentions of the id variable?

**Comment by Kevin-CB** [comment]
> have you retested after the latest changes?

Yes, last time I tested, it was on https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/678/commits/f18663b1faf723cab4fa7d10fb42a8e0fe90aa49

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (jglick) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:79]: Would have to look up some example, since I do not normally work on JS at all, but IIRC the point is that now that the behavior function is passed an element it would be able to use that argument to perform operations without a (unique) ID—either directly on the element in the argument, or on some sibling in the DOM.

In other words: when using inline `<script>` with a Jelly facet that might be included multiple times on a page, like the config snippet for a build step in a freestyle project config screen, you have the problem that

```xml
<body>
  <!-- first occurrence -->
  <div>
    <button onclick="handle()"/>
    <textarea id="spot"/>
    <script>
      function handle() {
        document.getElementById('spot').text = 'handled';
      }
    </script>
  </div>
</body>
```

will not work as soon as you add a second `<div>` with the same content: clicking either button will update the same textarea, perhaps the wrong one. `generateId` used to be used as a workaround:

```xml
<body>
  <!-- first occurrence -->
  <div>
    <j:set var="id" value="${h.generateId()}"/>
    <button onclick="handle()"/>
    <textarea id="spot-${id}"/>
    <script>
      function handle() {
        document.getElementById('spot-${id}').text = 'handled';
      }
    </script>
  </div>
</body>
```

When switching to a proper adjunct, you do not need this trick since

```xml
<body>
  <!-- first occurrence -->
  <div>
    <button class="mybutton"/>
    <textarea/>
    <st:adjunct …/>
  </div>
</body>
```

```js
function handle(button) {
  button.getSiblingByType('textarea').text = 'handled';
}
```

will work regardless of how many copies there are. (All code above is pseudocode, actual method names etc. are going to be a bit different.)

Of course in many cases the Jelly fragment could never be included more than once in a page because of what it is for, so whether a unique id was even needed here to begin with, I am unsure; it may have been included “just in case”.

**Review by jglick — APPROVED** [review]
Looks OK at a glance (relying on other reviewers for details & testing).

File: pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly
Comment (NeetigyaPod) [reply pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly:79]: Ohhhh makes sense. Thank you, will treat it as nit for future contributions in Jenkins.

**Pull request merged**
